src/coreclr/src/vm/assemblyloadcontext.cpp

```diff
@@ -14,3 +15,49 @@ HRESULT AssemblyLoadContext::GetBinderID(
     *pBinderId = reinterpret_cast<UINT_PTR>(this);
     return S_OK;
 }
+
+NativeImage *AssemblyLoadContext::LoadNativeImage(Module *componentModule, LPCUTF8 nativeImageName)
+{
+    STANDARD_VM_CONTRACT;
+
+#ifndef DACCESS_COMPILE
+    BaseDomain::LoadLockHolder lock(AppDomain::GetCurrentDomain());
+    AssemblyLoadContext *loadContext = componentModule->GetFile()->GetAssemblyLoadContext();
+    PTR_LoaderAllocator moduleLoaderAllocator = componentModule->GetLoaderAllocator();
+
+    int nativeImageCount = m_nativeImages.GetCount();
+    for (int nativeImageIndex = 0; nativeImageIndex < nativeImageCount; nativeImageIndex++)
+    {
+        NativeImage *nativeImage = m_nativeImages[nativeImageIndex];
+        if (nativeImage->Matches(nativeImageName))
+        {
+            return nativeImage;
+        }
+    }
+    
+    SString path = componentModule->GetPath();
+    SString::Iterator lastPathSeparatorIter = path.End();
+    size_t pathDirLength = 0;
+    if (PEAssembly::FindLastPathSeparator(path, lastPathSeparatorIter))
+    {
+        pathDirLength = (lastPathSeparatorIter - path.Begin()) + 1;
+    }
+
+    SString compositeImageFileName(SString::Utf8, nativeImageName);
+    SString fullPath;
+    fullPath.Set(path, path.Begin(), (COUNT_T)pathDirLength);
+    fullPath += compositeImageFileName;
+
+    PTR_PEImage peImage = PEImage::OpenImage(fullPath);
+    PTR_PEFile peFile = PEFile::Open(peImage);
+
+    NativeImage *nativeImage = NativeImage::Open(peFile, peImage->GetLoadedLayout(), nativeImageName, moduleLoaderAllocator);
+    if (nativeImage != NULL)
+    {
+        m_nativeImages.Append(nativeImage);
+        return nativeImage;
+    }
+#endif
```

Comment:
All of this logic feels temporary until we integrate all of this with the hosting surface and such. From my understanding, you've been putting together a system where we place multiple IL "PE" images into a single large file, and this is getting at the PE image of the actual wrapping large file. I don't understand why this needs to disk, instead of pulling the PEImage layout which is already loaded, somewhere. At the very least, this method needs a comment describing its purpose and description of how it attaches to the load flow of the runtime.

Reply:
Hmm, I'm not sure if I understand your comment. At this point I'm not trying to put "multiple IL PE images into a single large file" - as suggested by JanK and described in the readytorun-format.md design doc, I'm currently working on implementing "composite R2R with standalone MSIL" which means that a number of MSIL assemblies (the app, the framework, whatever) continues to reside in individual files on disk, they just share the "native code cache" represented by the compiled methods in the composite R2R file. I can easily improve the code comments or other documentation but I guess I first need to make sure the design is communicated clearly enough so that it makes sense at least to experts like you.

Reply:
Well, my comment is that in the long run IMO, there should not be *any* logic that is aware of file paths in the core engine of the runtime. Obviously as you put together the feature you have found that you need something here, but we will need a plan to remove all of that to an external, configurable bit of logic like the host.

As for the details around which variant of composite image you are making, the logic as written is extremely similar to what I would expect for the composite image model with embedded MSIL PE files. Its just lacking a discovery process for the embedded MSIL concept. I simply missed the distinction on production. (I blame going on vacation for a week.)

Reply:
Thanks for clarification, that makes sense to me. The decision to start with standalone MSIL is purely pragmatic as that requires fewer changes in the runtime exactly due to the fact that no changes are needed to MSIL discovery. I don't yet have a very clear idea as to whether the support for composite R2R with standalone MSIL is purely temporary or will remain as a longer-term supported scenario, I agree that the owner composite executable is a bit of an opportunistic hack rather than something backed by long-term strategic thinking.